File: integrations/mobile/android/Ads1299Packet.kt
package ads1299.core

data class Ads1299Packet(
    val version: Int,
    val flags: Int,
    val sequence: Long,
    val timestampUs: Long,
    val status: ByteArray,
    val channels: IntArray,
    val crc: Int,
    val raw: ByteArray,
)

object Ads1299Protocol {
    const val PACKET_SIZE = 49
    const val CHANNELS = 8
    const val VERSION = 0
    const val SYNC0 = 0xA5
    const val SYNC1 = 0x5A
    private const val CRC_OFFSET = 47

    fun crc16CcittFalse(data: ByteArray, offset: Int = 0, length: Int = data.size): Int {
        require(offset >= 0 && length >= 0 && offset + length <= data.size)
        var crc = 0xFFFF
        for (i in offset until offset + length) {
            crc = crc xor ((data[i].toInt() and 0xFF) shl 8)
            repeat(8) {
                crc = if ((crc and 0x8000) != 0) {
                    ((crc shl 1) xor 0x1021) and 0xFFFF
                } else {
                    (crc shl 1) and 0xFFFF
                }
            }
        }
        return crc
    }

    fun decode(raw: ByteArray): Ads1299Packet {
        require(raw.size == PACKET_SIZE) { "packet size ${raw.size} != $PACKET_SIZE" }
        require(u8(raw[0]) == SYNC0 && u8(raw[1]) == SYNC1) { "bad sync" }
        require(u8(raw[2]) == VERSION) { "unsupported version ${u8(raw[2])}" }

        val expected = crc16CcittFalse(raw, 2, CRC_OFFSET - 2)
        val actual = u16le(raw, CRC_OFFSET)
        require(actual == expected) {
            "CRC mismatch wire=0x${actual.toString(16)} expected=0x${expected.toString(16)}"
        }

        val status = raw.copyOfRange(12, 15)
        val channels = IntArray(CHANNELS) { ch -> i32le(raw, 15 + ch * 4) }

        return Ads1299Packet(
            version = u8(raw[2]),
            flags = u8(raw[3]),
            sequence = u32le(raw, 4),
            timestampUs = u32le(raw, 8),
            status = status,
            channels = channels,
            crc = actual,
            raw = raw.copyOf(),
        )
    }

    fun codeToMicrovolts(code: Int, vrefVolts: Double, gain: Int): Double {
        require(vrefVolts > 0.0)
        require(gain > 0)
        return code.toDouble() * vrefVolts / (gain.toDouble() * 8388607.0) * 1_000_000.0
    }

    private fun u8(value: Byte): Int = value.toInt() and 0xFF

    private fun u16le(data: ByteArray, offset: Int): Int =
        u8(data[offset]) or (u8(data[offset + 1]) shl 8)

    private fun u32le(data: ByteArray, offset: Int): Long =
        (u8(data[offset]).toLong()) or
            (u8(data[offset + 1]).toLong() shl 8) or
            (u8(data[offset + 2]).toLong() shl 16) or
            (u8(data[offset + 3]).toLong() shl 24)

    private fun i32le(data: ByteArray, offset: Int): Int =
        u8(data[offset]) or
            (u8(data[offset + 1]) shl 8) or
            (u8(data[offset + 2]) shl 16) or
            (u8(data[offset + 3]) shl 24)
}

File: integrations/mobile/ios_ipados/Ads1299Packet.swift
import Foundation

public struct Ads1299Packet: Sendable {
    public let version: UInt8
    public let flags: UInt8
    public let sequence: UInt32
    public let timestampUs: UInt32
    public let status: [UInt8]
    public let channels: [Int32]
    public let crc: UInt16
    public let raw: Data
}

public enum Ads1299ProtocolError: Error, CustomStringConvertible {
    case size(Int)
    case sync
    case version(UInt8)
    case crc(expected: UInt16, actual: UInt16)

    public var description: String {
        switch self {
        case .size(let value): return "packet size \(value) != 49"
        case .sync: return "bad ADS1299 packet sync"
        case .version(let value): return "unsupported protocol version \(value)"
        case .crc(let expected, let actual):
            return String(format: "CRC mismatch wire=0x%04X expected=0x%04X", actual, expected)
        }
    }
}

public enum Ads1299Protocol {
    public static let packetSize = 49
    public static let channels = 8
    public static let version: UInt8 = 0
    public static let sync0: UInt8 = 0xA5
    public static let sync1: UInt8 = 0x5A
    private static let crcOffset = 47

    public static func crc16CcittFalse(_ bytes: Data) -> UInt16 {
        var crc: UInt16 = 0xFFFF
        for byte in bytes {
            crc ^= UInt16(byte) << 8
            for _ in 0..<8 {
                if (crc & 0x8000) != 0 {
                    crc = (crc << 1) ^ 0x1021
                } else {
                    crc <<= 1
                }
            }
        }
        return crc
    }

    public static func decode(_ raw: Data) throws -> Ads1299Packet {
        guard raw.count == packetSize else { throw Ads1299ProtocolError.size(raw.count) }
        guard raw[0] == sync0 && raw[1] == sync1 else { throw Ads1299ProtocolError.sync }
        guard raw[2] == version else { throw Ads1299ProtocolError.version(raw[2]) }

        let expected = crc16CcittFalse(raw.subdata(in: 2..<crcOffset))
        let actual = u16le(raw, crcOffset)
        guard actual == expected else {
            throw Ads1299ProtocolError.crc(expected: expected, actual: actual)
        }

        let status = Array(raw[12..<15])
        var channelValues: [Int32] = []
        channelValues.reserveCapacity(channels)
        for channel in 0..<channels {
            channelValues.append(i32le(raw, 15 + channel * 4))
        }

        return Ads1299Packet(
            version: raw[2],
            flags: raw[3],
            sequence: u32le(raw, 4),
            timestampUs: u32le(raw, 8),
            status: status,
            channels: channelValues,
            crc: actual,
            raw: raw
        )
    }

    public static func codeToMicrovolts(_ code: Int32, vrefVolts: Double, gain: UInt) -> Double {
        guard vrefVolts > 0, gain > 0 else { return 0 }
        return Double(code) * vrefVolts / (Double(gain) * 8_388_607.0) * 1_000_000.0
    }

    private static func u16le(_ data: Data, _ offset: Int) -> UInt16 {
        UInt16(data[offset]) | (UInt16(data[offset + 1]) << 8)
    }

    private static func u32le(_ data: Data, _ offset: Int) -> UInt32 {
        UInt32(data[offset]) |
            (UInt32(data[offset + 1]) << 8) |
            (UInt32(data[offset + 2]) << 16) |
            (UInt32(data[offset + 3]) << 24)
    }

    private static func i32le(_ data: Data, _ offset: Int) -> Int32 {
        Int32(bitPattern: u32le(data, offset))
    }
}
